src/HolSmt/selftest.sml

```diff
@@ -551,11 +548,6 @@ in
     (``(x:real) < x``, [sat_CVC, sat_YO, sat_Z3, sat_Z3p]),
     (``(x:real) < y ==> 42 * x < 42 * y``, [thm_AUTO, thm_CVC, thm_YO, thm_Z3, thm_Z3p]),
 
-    (``0n <= 1n``, [thm_AUTO, thm_YO]),
-    (``1n <= 0n``, [sat_CVC, sat_YO, sat_Z3, sat_Z3p]),
-    (``(x:num) <= x``, [thm_AUTO, thm_YO]),
-    (``(x:num) <= y ==> 2 * x <= 2 * y``, [thm_AUTO, thm_YO]),
```

Comment:
Hi. Is this problem (`(x:num) <= y ==> 2 * x <= 2 * y`) supposed to be solvable only by Yices and not others?   I wonder why you have removed the above 4 cases, although the 2nd one (`1n <= 0m`) seems false.

Reply:
Right, so there are a few things going on here:

>  I wonder why you have removed the above 4 cases

So, the 4 test cases are being removed because the same test cases are already being performed above:
* The original test cases: https://github.com/HOL-Theorem-Prover/HOL/blob/5eb0e7b85e5c8789be46b9ff2a67d209d3981349/src/HolSmt/selftest.sml#L486-L489
* The duplicate ones: https://github.com/HOL-Theorem-Prover/HOL/blob/5eb0e7b85e5c8789be46b9ff2a67d209d3981349/src/HolSmt/selftest.sml#L554-L557

The duplicate ones are even in the wrong section (reals instead of nums). I suspect the duplicate ones might have been copy/pasted from the nums for reference, but someone forgot to delete them afterwards.

> Is this problem `((x:num) <= y ==> 2 * x <= 2 * y)` supposed to be solvable only by Yices and not others?

Currently, yes. The HolSmt support for Yices 1 uses a different translation which supports natural numbers. However, SMT-LIB doesn't support natural numbers, only integers, so we currently translate the natural numbers into uninterpreted constants.

It is possible to translate the natural numbers to the integers in an equivalence-preserving way (like it was done in Isabelle or perhaps F*) but we don't do this yet. I intend to fix this in the future (if nobody else does it first), since this is an extremely important use-case for me.

> the 2nd one (1n <= 0m) seems false.

It is indeed false. The HolSmt tests include tests for true statements (i.e. theorems) and false ones. When the SMT solvers can prove that the goal is false, they return `sat` instead of `unsat` (hence why that test case is tested with `sat_CVC`, `sat_Z3` and `sat_Z3p` rather than `thm_CVC`, `thm_Z3`, and `thm_Z3p`). In this case, it's easy for them to prove false statements about the natural numbers because they are allowed to assign whatever they want to uninterpreted constants, thus it's easy to assign something that makes the statement false.

This means that when we are forced to use uninterpreted constants or functions (which still happens a lot due to HolSmt limitations), a `sat` answer may not always imply that the goal is false, but an `unsat` answer always implies that the goal is true.

Hopefully, in the future we won't need to rely on uninterpreted symbols so much, making a `sat` answer a much more reliable indicator that the goal is false.

In my to-do list, there's also a task which consists in clarifying this to the HolSmt user: when the SMT solver replies `sat`, I'd like HolSmt to indicate to the user whether the `sat` result may be spurious or not. And also, I'd like to make it clear that when a non-spurious `sat` result is encountered, it indicates that the goal is false.